hooks/hookexecution/outcome.go

```diff
@@ -0,0 +1,77 @@
+package hookexecution
+
+import (
+	"time"
+
+	"github.com/prebid/prebid-server/hooks/hookanalytics"
+	"github.com/prebid/prebid-server/hooks/hookstage"
+)
+
+type Status string
+
+const (
+	StatusSuccess          Status = "success"
+	StatusTimeout          Status = "timeout"
+	StatusFailure          Status = "failure"           // expected module-side failure occurred during hook execution
+	StatusExecutionFailure Status = "execution_failure" // unexpected failure occurred during hook execution
+)
+
+type Action string
+
+const (
+	ActionUpdate Action = "update"
+	ActionReject Action = "reject"
+	ActionNOP    Action = "no_action"
+)
+
+type Messages map[string]map[string][]string // Messages in format: {"module": {"hook": ["msg1", "msg2"]}}
+
+type ModulesOutcome struct {
+	Errors   Messages     `json:"errors"`
+	Warnings Messages     `json:"warnings"`
+	Trace    TraceOutcome `json:"trace"`
+}
+
+type TraceOutcome struct {
+	ExecutionTime
+	Stages []Stage `json:"stages"`
+}
+
+type Stage struct {
+	ExecutionTime
+	Stage    string         `json:"stage"`
+	Outcomes []StageOutcome `json:"outcomes"`
+}
+
+type StageOutcome struct {
+	ExecutionTime
```

Comment:
There is a different calculation logic of the execution time for different outcome structures:
1. `GroupOutcome.ExecutionTime` is set to the longest `ExecutionTime` of its children, since hooks in a single group are executed concurrently.
2. `StageOutcome.ExecutionTime` uses the sum of the `ExecutionTime` of all its groups, because groups are executed serially.
3. `Stage.ExecutionTime` - is the longest `ExecutionTime` of its children. The bidder requests are sent concurrently.
4. And `TraceOutcome.ExecutionTime` - is the sum of `ExecutionTime` of all stages.

Reply:
Yes, you are right. The `bidder-request` and `raw-bidder-response` are the only stages that are executed multiple times (for each bidder). So these stages have multiple stage outcomes, thus the `Stage.ExecutionTime` field will be set to the longest `ExecutionTime` of its children.